infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServerQuorumReplicationDelegate.java

```diff
@@ -0,0 +1,152 @@
+                mode == WriteMode.QUORUM_PHASE2 || mode == WriteMode.QUORUM_STEADY;
+    }
+
+    boolean isFillHoleMessageRecognized(CorfuPayloadMsg<FillHoleRequest> msg) {
+        FillHoleMode mode = msg.getPayload().getFillHoleMode();
+        return mode == FillHoleMode.QUORUM_FORCE_OVERWRITE || mode == FillHoleMode.QUORUM_PHASE1 ||
+                mode == FillHoleMode.QUORUM_PHASE2 || mode == FillHoleMode.QUORUM_STEADY;
+    }
+
+
+
+
+
+    void write(CorfuPayloadMsg<WriteRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
+        log.debug("log quorum write: global: {}, streams: {}, backpointers: {}", msg
+                        .getPayload().getGlobalAddress(),
+                msg.getPayload().getStreamAddresses(), msg.getPayload().getData().getBackpointerMap());
+        // clear any commit record (or set initially to false).
+        msg.getPayload().clearCommit();
+        LogAddress addr = new LogAddress(msg.getPayload().getGlobalAddress(), null);
+        try (PrefixBasedLocalLocks.AutoCloseableLock quorum = locks.acquireWriteLock(addr.getAddress())) {
+            WriteMode mode = msg.getPayload().getWriteMode();
+            if (mode==WriteMode.QUORUM_FORCE_OVERWRITE) {
+                server.dataCache.put(addr, msg.getPayload().getData());
+                r.sendResponse(ctx, msg, CorfuMsgType.WRITE_OK.msg());
+                return;
+            }
+            long msgRank = msg.getPayload().getData().getRank();
+            LogData previous = server.dataCache.get(addr);
+            if (previous != null) {
+                if (!previous.getType().isProposal()) {
+                    r.sendResponse(ctx, msg, CorfuMsgType.ERROR_OVERWRITE.msg()); // no overwrite permitted
+                    return;
+                }
+                boolean phase1Matches = previous.getType().isData() && Arrays.equals(previous.getData(), msg.getPayload().getData().getData());
+                if (mode==WriteMode.QUORUM_PHASE1) { // message from phase 1
+                    if (phase1Matches) {
+                        r.sendResponse(ctx, msg, CorfuMsgType.WRITE_OK.msg()); // be idempotent
+                        return;
+                    }
+                    if (previous.getRank()>=msgRank) {
+                        r.sendResponse(ctx, msg, CorfuMsgType.ERROR_RANK.msg());
+                        return;
+                    }
+                    server.dataCache.put(addr, msg.getPayload().getData()); // for phase 1 - write only to the memory cache
+                    return;
+                } else { // Phase 2 or regular write
+                    if (previous.getRank()>msgRank || (previous.getRank()==msgRank && !phase1Matches)) {
+                        r.sendResponse(ctx, msg, CorfuMsgType.ERROR_RANK.msg());
+                        return;
+                    }
+                }
+            }
+            server.dataCache.put(addr, msg.getPayload().getData()); // so the regular, persistent write can happen
```

Comment:
The code would be more readable if we handle "if (previous == null)" case first, and return. 

the rest of the code is the "Otherwise" case (previous != null), one less indent,